### PR TITLE
Add RTL support

### LINKED ISSUE #1976: [IC3] Request: RTL support and localization for Hebrew, Urdu, and Arabic (+ info on localization contributions)
> VSTS #1663019

[edited by @corinagum)

Feature request: RTL support for languages like Hebrew, Urdu, and Arabic, as well as localization support for Hindi and the previously listed languages.
----

[Question]
is it possible speech to text of google is implement in bot frame work V4 like hindi and urdu and arabic?

### LINKED ISSUE #2902: RTL: Padding around typing indicator
## Screenshots

image

## Version

`corinagum/1976-RTL`

## Describe the bug

Typing indicator on RTL is not padded correctly.

## Steps to reproduce

1. Open Web Chat in RTL mode
1. Type `typing` to MockBot

## Expected behavior

> This is LTR.

image

## Additional context

[Bug]

### PR DESCRIPTION
**Pull request by corinagum**
Fixes #1976
Fixes #2902

## Changelog Entry
-  Resolves [#1976](https://github.com/microsoft/BotFramework-WebChat/issues/1976). Added RTL support with localization for Hebrew and Arabic, by [@corinagum](https://github.com/corinagum), in PR [#2890](https://github.com/microsoft/BotFramework-WebChat/pull/2890)

## Description
Implement the ability to display languages (such as Arabic, Urdu, etc) in Right-to-Left UI.

## What is Right to Left, Left to Right, and Bidirectional Support?
- Right-to-Left (RTL) is the UI implementation of displaying an app differently according to the language (human language, i.e. English or Arabic) of the app. For example, Arabic is a RTL language, and therefore an app in Arabic will ideally mirror<sup>*</sup> the English version of the UI to fit the directional flow of the language's script.
- Left-to-Right (LTR) follows the directional flow of languages like English
- Bidirectional support (or BiDi) allows for both RTL and LTR (Left-to-Right) support within a single app, depending on the language of the content being displayed.

(*) 'Mirror' in this article is not meant literally. In UI rules of BiDi, not every aspect or component will be mirrored, and the rules dictating those decisions will be discussed below.

## Why do we want Web Chat to support this?

- According to the [Ethnologue](https://www.ethnologue.com/guides/ethnologue200) of language studies, there are over 615 million speakers of RTL languages worldwide, which accounts for roughly 8% of the population
- There have been an increasing number of requests for Web Chat to support RTL or RTL languages. Below are some of those requests AND community-provided pull requests related to RTL languages:
   - [x] [#1976](https://github.com/microsoft/BotFramework-WebChat/issues/1976) - external participants indicating a desire for this feature
   - [x] [#2369](https://github.com/microsoft/BotFramework-WebChat/pull/2369) - `ar-AR`
        - Note: ar-AR is not an official language code, so this one will not be implemented.
   - [x] [#1548](https://github.com/microsoft/BotFramework-WebChat/pull/1548) `he-IL` 
      - stale implementation of RTL from an active Web Chat user - this will be used and credited as a starting block for our changes
   - [x] [#2477](https://github.com/microsoft/BotFramework-WebChat/pull/2477) `ar-JO`
   - [x] [#2593](https://github.com/microsoft/BotFramework-WebChat/pull/2593) `ar-EG`

   
### Pros
- Fits very neatly into Microsoft's mission of enabling the planet to achieve more
- Implementation time and difficulty appears relatively low
- We have a number of requests from customers that would like this feature
- Some of the work has been done already in the RTL PR [#2369](https://github.com/microsoft/BotFramework-WebChat/pull/2369)

### Cons
- Until ~Adaptive Cards and react-film~ `video` and `audio` have RTL support, Web Chat's RTL experience will not be fully functional
   - ~However, I believe that Web Chat having RTL support will encourage the Adaptive Cards team to also implement the feature~
- This is not a feature that will benefit 'most users' of Web Chat (80:20 rule)

### Web Chat dependencies that need RTL Support: 
1. HTML `<video>` and `<audio>` (not currently supported) 
1. YouTube & vimeo embed (not currently supported)
- [x] Adaptive Cards (UNOFFICIALLY supported) 
   - [Request](https://github.com/microsoft/AdaptiveCards/issues/3230) 
   - [Proposal](https://github.com/microsoft/AdaptiveCards/issues/2139)
   - Adaptive Cards team tentatively has RTL support proposed for AC 1.3 (Build 2020), but this is not finalized and may be bumped, as of 10/25/1019
1. [x] Markdown (supported) - works as expected when <body> dir attribute is set to 'rtl'
1. [x] [React Film](https://github.com/spyip/react-film) (supported) 
	- [Request](https://github.com/spyip/react-film/issues/31)
	
### Order of detection for script direction:
- [x] Allow Web Chat implementer to manually set prop (e.g. `dir: auto` (default)) to: `ltr` or `rtl` in `renderWebChat` props
- [x] Detect RTL settings based off of app's language

### Implementation steps as they apply to Web Chat
- [x] Set Web Chat prop indicating app's direction, based off data from "Order of detection for script direction" - this property will affect the entirety of Web Chat
- [x] Set Web Chat's ~root container's~ Basic Transcript and Basic Sendbox dir attribute depending on what was detected in 'Order of detection for script direction'
- [x] Flip right and left padding/spacing properties as needed 
- [x] Flip / use rtl versions of icons as needed
	a. Icons showing direction need to be flipped
	b. Icons that show 'right hand use' do not need to be flipped
- [x] Adopt localization from stale PRs listed above
- [x] Add `useDirection` documentation to `HOOKS.md`
- [x] Update LOCALIZATION.md
- [x] Test all browsers

### Icons/animations that need to be mirrored
   - [x] Airplane (signifying moving forward or proceeding to next part of conversation) 
   - [x] Typing indicator animation (follow the directional flow of the script)
   - [x] Carousel - react-film has been updated with a `dir` prop and/or automatic direction detection (See dependencies section above)
   - ~Spinner animation~ - I believe that spinners, going clockwise, are meant to represent the passage of time, and therefore should NOT be mirrored. Anecdotal and completely non-scientific information would point to a spinner remaining clockwise in RTL UI. No official-looking sources found on this topic
   - ~Sound slider SHOULD be mirrored - Audio / Video~ see dependencies above
   - ~Media CONTROLS SHOULD NOT be mirrored~ HTML5 does not support rtl for `<video>` or `<audio>` see dependencies above
   
### Situations to test
- ~HTML Videos~ - should be rendered by browser, but sound bar should be flipped
- ~Audio~ - should be rendered by browser, but sound bar should be flipped
- ~YouTube & vimeo~ - test that layout is correct
- [x] Adaptive Cards
- [x] Text activity - alignment
- [x] Switching languages between LTR and RTL
- [x] Download attachment - alignment
- [x] Connectivity Status component (icon Location and alignment)
- [x] Bubble nubs
- [x] `react-film`
   - [x] SuggestedActions
   - [x] CarouselFilmStrip
   - [x] Carousel Layout
- [x] ScrollToEnd button
- [x] Timestamp - number / date order / alignment
- [x] Send status component / Send failed retry
- [x] Markdown content - alignment
- [x] Samples - alignment
- [x] Verify all card (legacy) formats work with RTL
- [x] To IC3 team: Ask for card support priority
- [x] Avatar / Initials
- [ ] Ensure that Hosted Web Chat is able to specify RTL/Language like CDN and npm
- [ ] Have RTL Language speaker sign off with RTL Language bot - Test as many browsers as possible
- [x] RTL documentation

### Tests added to Mock Bot (still waiting on Arabic strings)
Mock Bot            
- [x] Greeting card 
- [x] Carousel      
- [x] File attachment
- [x] Markdown     

Offline Mock Bot            
- [x] Greeting card 
- [x] Carousel      
- ~[ ]~ File attachment
- [x] Markdown   

### Known internal issues
- [x] Fixed: <ScreenReaderText> component added offscreen `<div>` which created erroneous scrollbar in RTL mode

### Web features we will not use / Non-universal browser support
- `text-align`: 'start/end' - Not supported in Edge (or Internet Explorer)
- `match-parent` is only supported by Chrome and Firefox
- `<bdo>`

### Documentation / Guidance / Data
- https://hacks.mozilla.org/2015/09/building-rtl-aware-web-apps-and-websites-part-1/
- https://hacks.mozilla.org/2015/10/building-rtl-aware-web-apps-websites-part-2/
- https://material.io/design/usability/bidirectionality.html#mirroring-layout
- https://www.ethnologue.com/guides/ethnologue200

- [ ] Tests TBD
- [x] RTL sample added

### CODE REVIEW

File: packages/component/src/Activity/Bubble.js [hunk 1]
```diff
@@ -85,18 +82,20 @@ const Bubble = ({ 'aria-hidden': ariaHidden, children, className, fromUser, nub
...
       className={classNames(
         ROOT_CSS + '',
+        direction === 'rtl' ? 'webchat__bubble--rtl' : '',
         bubbleStyleSet + '',
         { 'from-user': fromUser, webchat__bubble_has_nub: nub },
         className + '' || ''
       )}
     >
-      <div className="webchat__bubble__content">{children}</div>
+      <div className={classNames('webchat__bubble__content')}>{children}</div>
```
Comment (corinagum) [inline packages/component/src/Activity/Bubble.js]: clean up

File: packages/component/src/Activity/Bubble.js [hunk 2]
```diff
@@ -42,7 +39,7 @@ function acuteNubSVG(nubSize, strokeWidth, side, upSideDown = false) {
 
   return (
     <svg
-      className="webchat__bubble__nub"
+      className={'webchat__bubble__nub'}
```
Comment (corinagum) [inline packages/component/src/Activity/Bubble.js]: clean up

File: packages/component/src/Styles/StyleSet/Bubble.js [hunk 3]
```diff
@@ -43,9 +43,8 @@ export default function createBubbleStyle({
 
     '&:not(.from-user)': {
       '&.webchat__bubble_has_nub': {
-        '& > .webchat__bubble__content': bubbleNubSize ? { marginLeft: paddingRegular } : {}
+        // '& > .webchat__bubble__content': bubbleNubSize ? { marginLeft: paddingRegular } : {}
```
Comment (corinagum) [inline packages/component/src/Styles/StyleSet/Bubble.js]: clean

File: packages/component/src/Attachment/FileContent.js [hunk 4]
```diff
@@ -25,6 +26,8 @@ const ROOT_CSS = css({
 });
 
 const FileContentBadge = ({ downloadIcon, fileName, size }) => {
+  const [direction] = useDirection();
+  console.log(direction);
```
Comment (corinagum) [inline packages/component/src/Attachment/FileContent.js]: clean up

File: packages/component/src/Activity/Bubble.js [hunk 5]
```diff
@@ -14,10 +15,6 @@ const ROOT_CSS = css({
   '& > .webchat__bubble__content': {
     // This is for hiding content outside of the bubble, for example, content outside of border radius
     overflow: 'hidden'
-  },
-
-  '& > .webchat__bubble__nub': {
```
Comment (corinagum) [inline packages/component/src/Activity/Bubble.js]: deleted dupe styling

**Comment by compulim** [comment]
image

image

**Comment by corinagum** [comment]
image

image

@compulim thanks, got it fixed :)

File: packages/component/src/Composer.js [hunk 6]
```diff
@@ -126,6 +126,7 @@ const Composer = ({
   attachmentRenderer,
   cardActionMiddleware,
   children,
+  dir,
```
Comment (compulim) [inline packages/component/src/Composer.js]: I think we should massage the input in `<Composer>` to make other code simpler..

For example, if the user send `<Composer dir="xyz">`, we should send `"auto"` to the context.

File: packages/component/src/SendBox/Assets/SendIcon.js [hunk 7]
```diff
@@ -1,9 +1,20 @@
...
-const SendIcon = () => (
-  <svg height={28} viewBox="0 0 45.7 33.8" width={28}>
-    <path clipRule="evenodd" d="M8.55 25.25l21.67-7.25H11zm2.41-9.47h19.26l-21.67-7.23zm-6 13l4-11.9L5 5l35.7 11.9z" />
-  </svg>
-);
+import useDirection from '../../hooks/useDirection';
+
+const SendIcon = () => {
+  const [direction] = useDirection();
+  const rtlScale = direction === 'rtl' ? css({ transform: 'scale(-1)' }) : '';
```
Comment (compulim) [inline packages/component/src/SendBox/Assets/SendIcon.js]: Use `const ROOT_CSS` (cache the result of `css(...)` call).

Also, `scale(-1, 1)` instead.

File: packages/component/src/SendBox/Assets/TypingAnimation.js [hunk 8]
```diff
@@ -1,16 +1,21 @@
...
 import React from 'react';
 
 import ScreenReaderText from '../../ScreenReaderText';
+import useDirection from '../../hooks/useDirection';
 import useStyleSet from '../../hooks/useStyleSet';
 
 const TypingAnimation = ({ 'aria-label': ariaLabel }) => {
   const [{ typingAnimation: typingAnimationStyleSet }] = useStyleSet();
+  const [direction] = useDirection();
+  const rtlScale = direction === 'rtl' ? css({ transform: 'scale(-1)' }) : '';
```
Comment (compulim) [inline packages/component/src/SendBox/Assets/TypingAnimation.js]: Cache it as `const ROOT_CSS = css({})` or something like `const RTL_SCALE_CSS`.

Also, `scale(-1, 1)` instead.

File: packages/component/src/SendBox/SuggestedAction.js [hunk 9]
```diff
@@ -50,7 +52,9 @@ const SuggestedAction = ({ 'aria-hidden': ariaHidden, buttonText, displayText, i
   return (
     <div aria-hidden={ariaHidden} className={classNames(suggestedActionStyleSet + '', SUGGESTED_ACTION_CSS + '')}>
       <button disabled={disabled} onClick={handleClick} type="button">
-        {image && <img src={image} />}
+        {image && (
+          <img className={classNames(direction === 'rtl' ? 'webchat__suggestedactions--rtl' : '')} src={image} />
```
Comment (compulim) [inline packages/component/src/SendBox/SuggestedAction.js]:
```suggestion
          <img className={classNames(direction === 'rtl' && 'webchat__suggestedactions--rtl')} src={image} />
```

File: packages/component/src/SendBox/SuggestedActions.js [hunk 10]
```diff
@@ -40,6 +41,7 @@ const connectSuggestedActions = (...selectors) =>
   );
 
 const SuggestedActions = ({ className, suggestedActions = [] }) => {
+  const [direction] = useDirection();
```
Comment (compulim) [inline packages/component/src/SendBox/SuggestedActions.js]: Sort. I believe `[direction]` should be ASCII-sorted below `[{ suggestedActionLayout, ...}]`.

File: packages/component/src/Styles/StyleSet/Bubble.js [hunk 11]
```diff
@@ -62,10 +58,30 @@ export default function createBubbleStyle({
...
+      },
+
+      '&.webchat__bubble--rtl': {
+        '&.webchat__bubble_has_nub': {
+          '& > .webchat__bubble__content': bubbleNubSize ? { marginRight: paddingRegular } : {}
+        },
+
+        '& > .webchat__bubble__nub': {
+          right: bubbleBorderWidth - bubbleNubSize + paddingRegular,
+          transform: 'scale(-1,1)'
```
Comment (compulim) [inline packages/component/src/Styles/StyleSet/Bubble.js]:
```suggestion
          transform: 'scale(-1, 1)'
```

File: packages/component/src/Styles/StyleSet/Bubble.js [hunk 12]
```diff
@@ -98,9 +110,28 @@ export default function createBubbleStyle({
...
+        '& > .webchat__bubble__nub': {
+          right: bubbleFromUserBorderWidth - bubbleFromUserNubSize + paddingRegular
+        }
+      },
+
+      '&.webchat__bubble--rtl': {
+        '&.webchat__bubble_has_nub': {
+          '& > .webchat__bubble__content': bubbleNubSize ? { marginLeft: paddingRegular } : {}
+        },
+        '& > .webchat__bubble__nub': {
```
Comment (compulim) [inline packages/component/src/Styles/StyleSet/Bubble.js]: Empty line before this? (Just being consistent with the block above.)

File: packages/component/src/Styles/StyleSet/SpinnerAnimation.js [hunk 13]
```diff
@@ -2,18 +2,31 @@ export default function createSpinnerAnimationStyle({
...
 
+  // since defaultStyleOptions has a predefined value for spinnerAnimationPadding by default, the console.warn will appear if both it and spinnerAnimationPaddingRight are set
+  if (spinnerAnimationPadding && spinnerAnimationPaddingRight) {
+    console.warn(
+      'botframework-webchat: "spinnerAnimationPaddingRight" is being deprecated and will be removed on or after 2021-2-1. Please use "spinnerAnimationPadding" instead.'
```
Comment (compulim) [inline packages/component/src/Styles/StyleSet/SpinnerAnimation.js]:
```suggestion
      'botframework-webchat: "spinnerAnimationPaddingRight" is being deprecated and will be removed on or after 2021-02-01. Please use "spinnerAnimationPadding" instead.'
```

I think it is 2022-02-01. 2 years + 2 major versions.

File: packages/component/src/Styles/StyleSet/SpinnerAnimation.js [hunk 14]
```diff
@@ -2,18 +2,31 @@ export default function createSpinnerAnimationStyle({
...
 
+  // since defaultStyleOptions has a predefined value for spinnerAnimationPadding by default, the console.warn will appear if both it and spinnerAnimationPaddingRight are set
+  if (spinnerAnimationPadding && spinnerAnimationPaddingRight) {
```
Comment (compulim) [inline packages/component/src/Styles/StyleSet/SpinnerAnimation.js]: Only show it once on first render of this type of component. For example,

```js
import React from 'react';

let deprecationNotesShown;

const SpinnerAnimation = () => {
  if (!deprecationNotesShown) {
    deprecationNotesShown = true;
    console.warn('botframework-webchat: ...');
  }

  return ();
};
```

File: packages/component/src/Styles/defaultStyleOptions.js [hunk 15]
```diff
@@ -147,13 +147,16 @@ const DEFAULT_OPTIONS = {
...
+  spinnerAnimationPadding: 12,
 
   enableUploadThumbnail: true,
   uploadThumbnailContentType: 'image/jpeg',
   uploadThumbnailHeight: 360,
   uploadThumbnailQuality: 0.6,
-  uploadThumbnailWidth: 720
+  uploadThumbnailWidth: 720,
+
+  // deprecated; will be removed on or after 2021-2-1
```
Comment (compulim) [inline packages/component/src/Styles/defaultStyleOptions.js]:
```suggestion
  // deprecated; will be removed on or after 2021-02-01
```

File: packages/component/src/hooks/useDirection.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import { useMemo } from 'react';
+
+import getRTLList from '../Localization/getRTLList';
+import useDirFromProps from './internal/useDirFromProps';
+import useLanguage from './useLanguage';
+
+function determineDirection(dir, language) {
+  if (dir === 'auto') {
```
Comment (compulim) [inline packages/component/src/hooks/useDirection.js]:
```js
if (dir !== 'auto') {
  return dir;
} else if (getRTLList().includes(language)) {
  return 'rtl';
}

return 'ltr';
```

File: packages/playground/src/App.js [hunk 17]
```diff
@@ -22,16 +22,16 @@ css.global('body', {
 
 const ROOT_CSS = css({
   height: '100%',
+  display: 'flex',
```
Comment (compulim) [inline packages/playground/src/App.js]: Sort.

File: samples/02.branding-styling-and-customization/g.change-locale/README.md [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+
+            const { useMemo, useState } = window.React;
+            const { createDirectLine, createStore, ReactWebChat } = window.WebChat;
+
+            const App = () => {
+               const [locale, setLocale] = useState(navigator.language);
+               const directLine = useMemo(() => createDirectLine({ token }), []);
+               const store = useMemo(
+                  () =>
+                     createStore({}, () => next => action => {
```
Comment (compulim) [inline samples/02.branding-styling-and-customization/g.change-locale/README.md]: Maybe improve this one using a button bar + `onhashchange` later.

File: samples/02.branding-styling-and-customization/g.change-locale/index.html [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        // In this demo, we are using Direct Line token from MockBot.
+        // Your client code must provide either a secret or a token to talk to your bot.
+        // Tokens are more secure. To learn about the differences between secrets and tokens
+        // and to understand the risks associated with using secrets, visit https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0
+
+        const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+        const { token } = await res.json();
+        const { ReactWebChat } = window.WebChat;
+
+        window.ReactDOM.render(
```
Comment (compulim) [inline samples/02.branding-styling-and-customization/g.change-locale/index.html]: Seems it is missing the implementation from `README.md`.

File: samples/02.branding-styling-and-customization/i.change-locale-and-direction/README.md [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+      </style>
+   </head>
+   <body>
+      <div id="webchat" role="main"></div>
+      <script>
+         (async function() {
+            const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
+            const { token } = await res.json();
+
+            window.WebChat.renderWebChat(
```
Comment (compulim) [inline samples/02.branding-styling-and-customization/i.change-locale-and-direction/README.md]: We could use a button bar + `onhashchange` later.

File: samples/02.branding-styling-and-customization/g.change-locale/index.html (hunk 19 above)
Comment (corinagum) [reply samples/02.branding-styling-and-customization/g.change-locale/index.html]: Not sure what you meant here. `README.md` and `index.html` code for `change-locale` are identical, minus comments

**Comment by cwhitten** [comment]
whoo hoo!

**Comment by geea-develop** [comment]
WOW Thanks! @corinagum cant wait to try it

**Pull request merged**